### PR TITLE
Add AsyncCommand from JT

### PR DESCRIPTION
**Pull request by jamesmontemagno**
Code from: https://johnthiriet.com/removing-async-void/#<3

### CODE REVIEW

**Comment by brminnick** [comment]
Hey James! 

I've been doing a lot of work lately with `AsyncCommand`, also inspired by John Thiret: https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices.MVVM/AsyncCommand.cs

I've been iterating & improving this implementation for about a year and it's all available on NuGet: https://www.nuget.org/packages/AsyncAwaitBestPractices.MVVM/

Here's a few tips/gotchyas I learned when implementing `AsyncCommand`:
- Subscribing to `CanExecuteChanged` can cause memory leaks for users
  - I recommend using a `WeakReference`, like I've implemented with [`WeakEventManager`](https://github.com/brminnick/AsyncAwaitBestPractices/blob/e768a96e2ceddc7644eee54d60cbd9343ca857b7/Src/AsyncAwaitBestPractices/WeakEventManager/WeakEventManager.cs) to avoid introducing any memory leaks into a user's code base
- Validating the parameter passed into `ICommand.Execute(object parameter)` for `IAsyncCommand<T>`
  - It's valid for `T` to be `null`, but only if `T` is a Reference Type. E.g. For `var boolCommand = new AsyncCommand<bool>()`, `myCommand.Execute(null)` won't work.
  - I implemented [validation along with a specific `InvalidCommandParameterException`](https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices.MVVM/AsyncCommand.cs#L71-L79) to provide a more helpful error message to the user
- In `FireAndForgetSafeAsync `, use `.ConfigureAwait(false)`, e.g. `await task.ConfigureAwait(false)`
  - This improves performance by preventing an unnecessary context switch once the Task has completed
  - I [implemented it as the default, but also give the user the option to use `.ConfigureAwait(true)` if the need it](https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices/SafeFireAndForgetExtensions.cs#L46-L60)
- In `FireAndForgetSafeAsync`, exception filters are more performant than catching everything first then checking if `handler` is null
   - If you check out the IL, `catch (Exception e) when (handler != null)` compiles down to more performant code than `catch (Exception e){ handler?.HandleError(ex); }

File: MvvmHelpers/Commands/Command.cs [hunk 1]
```diff
@@ -4,85 +4,85 @@
...
-    /// <summary>
-    /// Implementaiton of ICommand
-    /// </summary>
-    public class Command : ICommand
-    {
-        readonly Func<object, bool> canExecute;
-        readonly Action<object> execute;
-        readonly WeakEventManager weakEventManager = new WeakEventManager();
+	/// <summary>
+	/// Implementaiton of ICommand
```
Comment (johnthiriet) [inline MvvmHelpers/Commands/Command.cs]: Typo => Implementation

File: MvvmHelpers/Utils.cs [hunk 2]
```diff
@@ -5,81 +5,81 @@
...
+		}
 
-        internal static bool IsValidParameter<T>(object o)
-        {
-            bool valid;
-            if (o != null)
-            {
-                // The parameter isn't null, so we don't have to worry whether null is a valid option
-                valid = o is T;
+		internal static bool IsValidParameter<T>(object o)
```
Comment (johnthiriet) [inline MvvmHelpers/Utils.cs]: Maybe we could find a better name that actually indicates for which class we are checking if the parameter is a valid one.

File: MvvmHelpers.UnitTests/WeakEventManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+
+namespace MvvmHelpers.UnitTests
+{
+	[TestClass]
+	public class WeakEventManagerTests
+	{
+		static int count;
```
Comment (johnthiriet) [inline MvvmHelpers.UnitTests/WeakEventManagerTests.cs]: I usually use Xunit which runs tests randomly and in parallel therefore seeing `static` in a unit test class always triggers a warning in my head. The comment is here just in case the behavior is similar and could cause breaking tests.

File: MvvmHelpers/Exceptions/InvalidCommandParameterException.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+{
+	/// <summary>
+	/// Represents errors that occur during IAsyncCommand execution.
+	/// </summary>
+	public class InvalidCommandParameterException : Exception
+	{
+		/// <summary>
+		/// Initializes a new instance of the <see cref="T:MvvmHelpersInvalidCommandParameterException"/> class.
+		/// </summary>
+		/// <param name="excpectedType">Excpected parameter type for AsyncCommand.Execute.</param>
```
Comment (johnthiriet) [inline MvvmHelpers/Exceptions/InvalidCommandParameterException.cs]: Typo => `excpectedType` and `Excpected`

File: MvvmHelpers/Exceptions/InvalidCommandParameterException.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+		/// <param name="innerException">Inner Exception</param>
+		public InvalidCommandParameterException(Type excpectedType, Type actualType, Exception innerException) : base(CreateErrorMessage(excpectedType, actualType), innerException)
+		{
+
+		}
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="T:MvvmHelpers.InvalidCommandParameterException"/> class.
+		/// </summary>
+		/// <param name="excpectedType">Excpected parameter type for AsyncCommand.Execute.</param>
```
Comment (johnthiriet) [inline MvvmHelpers/Exceptions/InvalidCommandParameterException.cs]: Typo => `excpectedType` and `Excpected`

File: MvvmHelpers/Commands/CommandUtils.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+			var t = typeof(T);
+
+			// The parameter is null. Is T Nullable?
+			if (Nullable.GetUnderlyingType(t) != null)
+			{
+				return true;
+			}
+
+			// Not a Nullable, if it's a value type then null is not valid
+			valid = !t.GetTypeInfo().IsValueType;
```
Comment (brminnick) [inline MvvmHelpers/Commands/CommandUtils.cs]: Just FYI - `IsValueType` is only available in .NET Standard 2.0:  https://docs.microsoft.com/dotnet/api/system.type.isvaluetype.

It looks like MvvmHelpers' current minimum dependency is .NET Standard 1.0, so make sure to bump it up to .NET Standard 2.0.

MvvmHelpers also targets . NET 4.6.1, which is fine because `IsValueType` is supported in .NET v1.1.

File: MvvmHelpers/Utils.cs [hunk 7]
```diff
@@ -1,38 +1,54 @@
...
+
+#pragma warning disable RECS0165 // Asynchronous methods should return a Task instead of void
+		public static async void SafeFireAndForgetAsync(this Task task, Action<Exception> onException = null, bool continueOnCapturedContext = false)
+#pragma warning restore RECS0165 // Asynchronous methods should return a Task instead of void
+		{
+			try
+			{
+				await task.ConfigureAwait(continueOnCapturedContext);
+			}
+			catch (Exception ex)
```
Comment (brminnick) [inline MvvmHelpers/Utils.cs]: Careful - this is dangerous because it catches every exception thrown by the `Task`. 

I recommend not catching the Exception if the user has not implemented `onException` because it means that they don't want to gracefully handle the exception. I like to think of `onException` as the way the user implements a try/catch for `FireAndForgetAsync`; if they haven't implemented `onException`, don't catch any exceptions; if they have, catch 'em (all 😜).

```csharp
try
{
    await task.ConfigureAwait(continueOnCapturedContext);
}
catch (Exception ex)
{
    if(onException is null)
        throw;

    onException.Invoke(ex);
}
```

Same code, but optimized using Catch filters as recommended by the great Bill Wagner: https://github.com/brminnick/AsyncAwaitBestPractices/pull/8
```csharp
try
{
    await task.ConfigureAwait(continueOnCapturedContext);
}
catch (Exception ex) when (ex != null)
{
    onException.Invoke(ex);
}
```

File: MvvmHelpers/WeakEventManager.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+using static System.String;
+
+namespace MvvmHelpers
+{
+	public class WeakEventManager
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: I recommend making this class `internal`, because otherwise devs will start using it and you'll beholden to maintaining a `WeakEventManager` despite this library focusing on MVVM. I implemented it in [AsyncAwaitBestPractices](https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices/WeakEventManager/WeakEventManager.cs), and I wish I had kept it `internal` so that only `AsyncCommand` would use it. Here's why:

**Story Time**

I released [AsyncAwaitBestPracitces](https://www.nuget.org/packages/AsyncAwaitBestPractices/), with `WeakEventManager` public, and shortly after its release someone asked me for `WeakEventManager<T>` for `EventHandler<T>`- sure, no problem! 

But then someone pointed out that not every `event` is an `EventHandler` and that events can also be a `delegate` (like [`NotifyPropertyChanged`](https://docs.microsoft.com/dotnet/api/system.componentmodel.inotifypropertychanged.propertychanged?view=netframework-4.8)), so [I added support for `Delegate`](https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices/WeakEventManager/WeakEventManager.cs#L128).

But THEN someone pointed out that `Action` can also be used for events, and so I added support for [`event Action`](https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices/WeakEventManager/WeakEventManager.cs#L39). But this was easier said than done, because now I needed to add an overloaded method for `WeakEventManager.HandleEvent` because `Action` and `Action<T>` doesn't consume an object, so I also ended up creating a custom Exception to throw when a dev uses the incorrect overloaded method for `HandleEvent`: https://github.com/brminnick/AsyncAwaitBestPractices/blob/master/Src/AsyncAwaitBestPractices/WeakEventManager/EventManagerService.cs#L58

tl;dr - save yourself the headaches and make this `internal`

File: MvvmHelpers/WeakEventManager.cs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	{
+		readonly Dictionary<string, List<Subscription>> eventHandlers = new Dictionary<string, List<Subscription>>();
+
+		public void AddEventHandler<TEventArgs>(EventHandler<TEventArgs> handler, [CallerMemberName]string eventName = null)
+			where TEventArgs : EventArgs
+		{
+			if (IsNullOrEmpty(eventName))
+				throw new ArgumentNullException(nameof(eventName));
+
+			if (handler == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/WeakEventManager.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+			AddEventHandler(eventName, handler.Target, handler.GetMethodInfo());
+		}
+
+		public void AddEventHandler(EventHandler handler, [CallerMemberName]string eventName = null)
+		{
+			if (IsNullOrEmpty(eventName))
+				throw new ArgumentNullException(nameof(eventName));
+
+			if (handler == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/WeakEventManager.cs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+					if (isStatic)
+					{
+						// For a static method, we'll just pass null as the first parameter of MethodInfo.Invoke
+						toRaise.Add((null, subscription.Handler));
+						continue;
+					}
+
+					var subscriber = subscription.Subscriber.Target;
+
+					if (subscriber == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/WeakEventManager.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+			}
+		}
+
+		public void RemoveEventHandler<TEventArgs>(EventHandler<TEventArgs> handler, [CallerMemberName]string eventName = null)
+			where TEventArgs : EventArgs
+		{
+			if (IsNullOrEmpty(eventName))
+				throw new ArgumentNullException(nameof(eventName));
+
+			if (handler == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/WeakEventManager.cs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+			RemoveEventHandler(eventName, handler.Target, handler.GetMethodInfo());
+		}
+
+		public void RemoveEventHandler(EventHandler handler, [CallerMemberName]string eventName = null)
+		{
+			if (IsNullOrEmpty(eventName))
+				throw new ArgumentNullException(nameof(eventName));
+
+			if (handler == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/WeakEventManager.cs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+		void AddEventHandler(string eventName, object handlerTarget, MethodInfo methodInfo)
+		{
+			if (!eventHandlers.TryGetValue(eventName, out var targets))
+			{
+				targets = new List<Subscription>();
+				eventHandlers.Add(eventName, targets);
+			}
+
+			if (handlerTarget == null)
```
Comment (brminnick) [inline MvvmHelpers/WeakEventManager.cs]: Use `handler is null` to squeeze out some more performance benefits: https://www.gullberg.tk/blog/is-null-versus-null-in-c/

**tl;dr** Pattern Matching is faster than Equality comparison in the .NET IL

File: MvvmHelpers/Utils.cs [hunk 15]
```diff
@@ -1,38 +1,54 @@
...
+		/// </summary>
+		/// <returns>The task with timeout.</returns>
+		/// <param name="task">Task.</param>
+		/// <param name="timeout">Timeout Duration.</param>
+		/// <typeparam name="T">The 1st type parameter.</typeparam>
+		public static Task<T> WithTimeout<T>(this Task<T> task, TimeSpan timeout) =>
+			WithTimeout(task, (int)timeout.TotalMilliseconds);
+
+#pragma warning disable RECS0165 // Asynchronous methods should return a Task instead of void
+		public static async void SafeFireAndForgetAsync(this Task task, Action<Exception> onException = null, bool continueOnCapturedContext = false)
```
Comment (brminnick) [inline MvvmHelpers/Utils.cs]: I recommend removing the `Async` suffix because it may confuse devs who try to `await` `SafeFireAndForgetAsync` which cannot be done. Instead, change it to `SafeFireAndForget`.

File: MvvmHelpers/Commands/CommandUtils.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+		internal static bool IsValidCommandParameter<T>(object o)
+		{
+			bool valid;
+			if (o != null)
+			{
+				// The parameter isn't null, so we don't have to worry whether null is a valid option
+				valid = o is T;
+
+				if (!valid)
+					throw new InvalidCommandParameterException(typeof(T), null);
```
Comment (brminnick) [inline MvvmHelpers/Commands/CommandUtils.cs]: Passing in `null` to `new InvalidCommandParameterException(Type expectedType, Type actualType)` creates a confusing error message:

```
Unhandled Exception: MvvmHelpers.Exceptions.InvalidCommandParameterException: Invalid type for parameter. Expected Type: System.Int32, but received Type: 
   at MvvmHelpers.Commands.CommandUtils.IsValidCommandParameter[T](Object o)
```

Since you want to let the user know that they incorrectly passed in a `null`, you could create  another constructor for `InvalidCommandParameterException`:

```csharp
public InvalidCommandParameterException(Type expectedType) : base("Invalid type for parameter. Expected Type: {expectedType}, but received Type: null")
{

} 
```

File: MvvmHelpers/Commands/CommandUtils.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+			if (Nullable.GetUnderlyingType(t) != null)
+			{
+				return true;
+			}
+
+			// Not a Nullable, if it's a value type then null is not valid
+			valid = !t.GetTypeInfo().IsValueType;
+
+			if (!valid)
+				throw new InvalidCommandParameterException(typeof(T), null);
```
Comment (brminnick) [inline MvvmHelpers/Commands/CommandUtils.cs]: Passing in `null` to `new InvalidCommandParameterException(Type expectedType, Type actualType)` creates a confusing error message:

```
Unhandled Exception: MvvmHelpers.Exceptions.InvalidCommandParameterException: Invalid type for parameter. Expected Type: System.Int32, but received Type: 
   at MvvmHelpers.Commands.CommandUtils.IsValidCommandParameter[T](Object o)
```

I recommend creating  another constructor for `InvalidCommandParameterException`:

```csharp
public InvalidCommandParameterException(Type expectedType) : base("Invalid type for parameter. Expected Type: {expectedType}")
{

} 
```

File: MvvmHelpers/Exceptions/InvalidCommandParameterException.cs [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+		/// Initializes a new instance of the <see cref="T:MvvmHelpers.InvalidCommandParameterException"/> class.
+		/// </summary>
+		/// <param name="expectedType">Expected parameter type for AsyncCommand.Execute.</param>
+		/// <param name="actualType">Actual parameter type for AsyncCommand.Execute.</param>
+		public InvalidCommandParameterException(Type expectedType, Type actualType) : base(CreateErrorMessage(expectedType, actualType))
+		{
+
+		}
+
+		static string CreateErrorMessage(Type expectedType, Type actualType) =>
```
Comment (brminnick) [inline MvvmHelpers/Exceptions/InvalidCommandParameterException.cs]: Add two more constructors here to align with this feedback: https://github.com/jamesmontemagno/mvvm-helpers/pull/43#discussion_r315461010

```csharp
/// <summary>
/// Initializes a new instance of the <see cref="T:TaskExtensions.MVVM.InvalidCommandParameterException"/> class.
/// </summary>
/// <param name="excpectedType">Excpected parameter type for AsyncCommand.Execute.</param>
/// <param name="innerException">Inner Exception</param>
public InvalidCommandParameterException(Type excpectedType, Exception innerException) : base(CreateErrorMessage(excpectedType), innerException)
{

}

/// <summary>
/// Initializes a new instance of the <see cref="T:TaskExtensions.MVVM.InvalidCommandParameterException"/> class.
/// </summary>
/// <param name="excpectedType">Excpected parameter type for AsyncCommand.Execute.</param>
public InvalidCommandParameterException(Type excpectedType) : base(CreateErrorMessage(excpectedType))
{

}

static string CreateErrorMessage(Type excpectedType) => $"Invalid type for parameter. Expected Type {excpectedType}";
```
